File: packages/risk-core/src/budget-types.ts
/**
 * @file budget-types.ts
 * Core Budget, Route Cost Policy, and Cost-Aware Rate Limiter Type Definitions.
 */

export type HttpMethod = 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH' | '*';
export type FailureMode = 'allow' | 'allow_with_emergency_cap' | 'deny';
export type AuthRequirement = 'optional' | 'required' | 'admin_only';
export type CostGuardAction = 'ALLOW' | 'OBSERVE' | 'RATE_LIMIT' | 'REQUIRE_AUTH' | 'DENY';
export type CostGuardReasonCode =
  | 'WITHIN_BUDGET'
  | 'REQUEST_SHAPE_EXCEEDED'
  | 'COST_BUDGET_EXCEEDED'
  | 'LIMITER_UNAVAILABLE'
  | 'POLICY_NOT_FOUND'
  | 'POLICY_INVALID'
  | 'AUTH_REQUIRED'
  | 'FAIL_CLOSED'
  | 'FAIL_OPEN';

export interface RouteCostPolicy {
  method: HttpMethod;
  path: string;
  cost: number;
  authentication?: AuthRequirement;
  page_size_default?: number;
  page_size_max?: number;
  query_timeout_ms?: number;
  lock_timeout_ms?: number;
  max_data_points?: number;
  max_request_body_bytes?: number;
  max_response_bytes?: number;
  failure_mode?: FailureMode;
  shadow_mode?: boolean;
}

export interface RateLimitTier {
  name?: string;
  capacity: number;
  refill_tokens_per_minute: number;
  emergency_local_capacity: number;
}

export interface CostPolicyDefaults {
  cost: number;
  authentication: AuthRequirement;
  page_size_default: number;
  page_size_max: number;
  query_timeout_ms: number;
  lock_timeout_ms: number;
  max_data_points: number;
  max_request_body_bytes: number;
  max_response_bytes: number;
  failure_mode: FailureMode;
  shadow_mode: boolean;
}

export interface CostPolicyConfig {
  schema_version: string;
  policy_version: string;
  defaults: CostPolicyDefaults;
  rate_limit_tiers?: Record<string, RateLimitTier>;
  routes?: RouteCostPolicy[];
}

export interface PolicyIdentity {
  checksumSha256: string;
  displayChecksum: string;
}

export type VerifiedPrincipal =
  | {
      authenticated: true;
      tenantId?: string;
      accountId?: string;
      apiKeyId?: string;
      role?: string;
      tier?: string;
      authenticatedTier?: string;
      riskClass?: string;
    }
  | {
      authenticated: false;
      tenantId?: undefined;
      accountId?: undefined;
      apiKeyId?: undefined;
      role?: undefined;
      tier?: undefined;
      authenticatedTier?: undefined;
      riskClass?: string;
    };


export interface RequestCostContext {
  method: string;
  path: string;
  bodyBytes?: number;
  pageSize?: number;
  seriesCount?: number;
  timeBuckets?: number;
  principal?: VerifiedPrincipal;
  sessionId?: string;
  pseudonymousKey?: string;
  asn?: number;
  tenantId?: string; // Raw/unverified request header/query context
}

export interface CostGuardDecision {
  allowed: boolean;
  action: CostGuardAction;
  proposedAction: CostGuardAction;
  enforcedAction: CostGuardAction;
  violationDetected: boolean;
  reasonCode: CostGuardReasonCode;
  policyVersion: string;
  policyChecksum: string;
  displayChecksum: string;
  cost: number;
  budget?: {
    requestedCost: number;
    remainingCost: number;
    retryAfterSeconds: number;
  };
  degraded?: boolean;
  enforced?: boolean;
  message?: string;
}

export const BUDGET_SCOPES = [
  'global',
  'route',
  'tenant',
  'account',
  'authKey',
  'session',
  'network',
] as const;

export type BudgetScope = typeof BUDGET_SCOPES[number];

export const LEGACY_DEFAULT_BUDGET_SCOPES = [
  'route',
  'tenant',
  'account',
  'authKey',
  'session',
  'network',
] as const;

export type LegacyBudgetScope = typeof LEGACY_DEFAULT_BUDGET_SCOPES[number];

export interface ScopeBudgetConfig {
  capacity: number;
  refillRatePerSec: number;
}

export interface BudgetConsumeRequest {
  cost: number;
  routeKey: string;
  requestId?: string;
  idempotencyTtlSeconds?: number;
  globalKey?: string;
  tenantId?: string;
  accountId?: string;
  apiKeyId?: string;
  sessionId?: string;
  networkKey?: string;
  tier?: RateLimitTier;
  emergencyCapacity?: number;
  policy?: RouteCostPolicy;
  scopeBudgets?: {
    global?: ScopeBudgetConfig;
    route?: ScopeBudgetConfig;
    tenant?: ScopeBudgetConfig;
    account?: ScopeBudgetConfig;
    authKey?: ScopeBudgetConfig;
    session?: ScopeBudgetConfig;
    network?: ScopeBudgetConfig;
  };
}


export type RedisFailureMode =
  | 'fail-closed'
  | 'fail-open'
  | 'local-emergency';

export interface RedisFailurePolicy {
  mode: RedisFailureMode;
  emergencyRatio?: number;
  expectedReplicaCount?: number;
  riskLevel?: 'low' | 'medium' | 'high' | 'critical';
}

export interface BudgetConsumeResult {
  allowed: boolean;
  remainingCost: number;
  retryAfterSeconds: number;
  degraded: boolean;
  store: 'redis' | 'local-emergency' | 'bypass';
  consistency: 'distributed-atomic' | 'process-local' | 'none';
  reason:
    | 'ALLOWED'
    | 'QUOTA_EXCEEDED'
    | 'REDIS_UNAVAILABLE'
    | 'FAIL_CLOSED'
    | 'FAIL_OPEN'
    | 'INVALID_REQUEST';
}

export interface BudgetStore {
  consume(request: BudgetConsumeRequest): Promise<BudgetConsumeResult>;
}

export interface RedactedThreatEvent {
  signature: string;
  routeTemplate: string;
  asn: number;
  statusCode?: number;
  evidenceCode?: string;
  timestamp: number;
}

export interface ThreatAggregateRecord {
  windowStart: string;
  windowSeconds: number;
  signature: string;
  routeTemplate: string;
  asn: number;
  count: number;
  samples: Array<{ timestamp: number; statusCode?: number; evidenceCode?: string }>;
}

export interface ThreatAggregateStore {
  increment(event: RedactedThreatEvent): Promise<void>;
  flush?(): Promise<ThreatAggregateRecord[]>;
}

File: packages/sentinel-py/src/ameva_sentinel/core/budget_types.py
"""
Type definitions and Pydantic validation models for AMEVA-Sentinel Python SDK.
Strict validation preventing type coercion errors and enforcing schema parity.
"""

from typing import Literal, Optional, List, Dict, Any
from dataclasses import dataclass, field
from pydantic import BaseModel, ConfigDict, Field, StrictInt, StrictStr, StrictBool, model_validator

HttpMethod = Literal["GET", "POST", "PUT", "DELETE", "PATCH", "*"]
FailureMode = Literal["allow", "allow_with_emergency_cap", "deny"]
AuthRequirement = Literal["optional", "required", "admin_only"]
CostGuardAction = Literal["ALLOW", "OBSERVE", "RATE_LIMIT", "REQUIRE_AUTH", "DENY"]
CostGuardReasonCode = Literal[
    "WITHIN_BUDGET",
    "REQUEST_SHAPE_EXCEEDED",
    "COST_BUDGET_EXCEEDED",
    "LIMITER_UNAVAILABLE",
    "POLICY_NOT_FOUND",
    "POLICY_INVALID",
    "AUTH_REQUIRED",
]


class RouteCostPolicy(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)

    method: HttpMethod
    path: StrictStr = Field(..., min_length=1)
    cost: StrictInt = Field(..., gt=0)
    authentication: Optional[AuthRequirement] = None
    page_size_default: Optional[StrictInt] = Field(None, gt=0)
    page_size_max: Optional[StrictInt] = Field(None, gt=0)
    query_timeout_ms: Optional[StrictInt] = Field(None, ge=10, le=60000)
    lock_timeout_ms: Optional[StrictInt] = Field(None, ge=10, le=10000)
    max_data_points: Optional[StrictInt] = Field(None, gt=0)
    max_request_body_bytes: Optional[StrictInt] = Field(None, ge=1024, le=52428800)
    max_response_bytes: Optional[StrictInt] = Field(None, ge=1024, le=52428800)
    failure_mode: Optional[FailureMode] = None
    shadow_mode: Optional[StrictBool] = None

    @model_validator(mode="after")
    def validate_page_size_bounds(self) -> "RouteCostPolicy":
        if (
            self.page_size_default is not None
            and self.page_size_max is not None
            and self.page_size_default > self.page_size_max
        ):
            raise ValueError("page_size_default cannot exceed page_size_max")
        return self


class RateLimitTier(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)

    name: Optional[StrictStr] = None
    capacity: StrictInt = Field(..., gt=0)
    refill_tokens_per_minute: StrictInt = Field(..., gt=0)
    emergency_local_capacity: StrictInt = Field(..., gt=0)


class CostPolicyDefaults(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)

    cost: StrictInt = Field(..., gt=0)
    authentication: AuthRequirement = "optional"
    page_size_default: StrictInt = Field(25, gt=0)
    page_size_max: StrictInt = Field(50, gt=0)
    query_timeout_ms: StrictInt = Field(2000, ge=10, le=60000)
    lock_timeout_ms: StrictInt = Field(500, ge=10, le=10000)
    max_data_points: StrictInt = Field(5000, gt=0)
    max_request_body_bytes: StrictInt = Field(1048576, ge=1024, le=52428800)
    max_response_bytes: StrictInt = Field(1048576, ge=1024, le=52428800)
    failure_mode: FailureMode = "allow_with_emergency_cap"
    shadow_mode: StrictBool = True

    @model_validator(mode="after")
    def validate_page_size_bounds(self) -> "CostPolicyDefaults":
        if self.page_size_default > self.page_size_max:
            raise ValueError("defaults.page_size_default cannot exceed defaults.page_size_max")
        return self


class CostPolicyConfig(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)

    schema_version: Literal["1.0"] = "1.0"
    policy_version: StrictStr = Field(..., min_length=1)
    defaults: CostPolicyDefaults
    rate_limit_tiers: Dict[str, RateLimitTier] = Field(default_factory=dict)
    routes: List[RouteCostPolicy] = Field(default_factory=list)


@dataclass(frozen=True)
class PolicyIdentity:
    checksum_sha256: str
    display_checksum: str


@dataclass
class VerifiedPrincipal:
    authenticated: bool
    tenant_id: Optional[str] = None
    account_id: Optional[str] = None
    api_key_id: Optional[str] = None
    role: Optional[str] = None
    tier: Optional[str] = None
    authenticated_tier: Optional[str] = None
    risk_class: str = "anonymous_network"

    def __post_init__(self) -> None:
        if not self.authenticated:
            if any(
                val is not None
                for val in (self.tenant_id, self.account_id, self.api_key_id, self.role, self.tier, self.authenticated_tier)
            ):
                raise ValueError(
                    "Unauthenticated VerifiedPrincipal cannot contain identity fields "
                    "(tenant_id, account_id, api_key_id, role, tier/authenticated_tier must be None when authenticated=False)"
                )




@dataclass
class RequestCostContext:
    method: str
    path: str
    body_bytes: Optional[int] = None
    page_size: Optional[int] = None
    series_count: Optional[int] = None
    time_buckets: Optional[int] = None
    principal: Optional[VerifiedPrincipal] = None
    session_id: Optional[str] = None
    pseudonymous_key: Optional[str] = None
    asn: Optional[int] = None
    tenant_id: Optional[str] = None  # Unverified context identifier


@dataclass
class CostGuardBudgetInfo:
    requested_cost: int
    remaining_cost: int
    retry_after_seconds: int = 0


@dataclass
class CostGuardDecision:
    allowed: bool
    action: CostGuardAction
    proposed_action: CostGuardAction
    enforced_action: CostGuardAction
    violation_detected: bool
    reason_code: CostGuardReasonCode
    policy_version: str
    policy_checksum: str
    display_checksum: str
    cost: int
    budget: Optional[CostGuardBudgetInfo] = None
    degraded: bool = False
    enforced: bool = False
    message: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        return {
            "allowed": self.allowed,
            "action": self.action,
            "proposedAction": self.proposed_action,
            "enforcedAction": self.enforced_action,
            "violationDetected": self.violation_detected,
            "reasonCode": self.reason_code,
            "policyVersion": self.policy_version,
            "policyChecksum": self.policy_checksum,
            "displayChecksum": self.display_checksum,
            "cost": self.cost,
            "budget": {
                "requestedCost": self.budget.requested_cost,
                "remainingCost": self.budget.remaining_cost,
                "retryAfterSeconds": self.budget.retry_after_seconds,
            } if self.budget else None,
            "degraded": self.degraded,
            "enforced": self.enforced,
            "message": self.message,
        }


BUDGET_SCOPES: tuple[str, ...] = (
    "global",
    "route",
    "tenant",
    "account",
    "authKey",
    "session",
    "network",
)

LEGACY_DEFAULT_BUDGET_SCOPES: tuple[str, ...] = (
    "route",
    "tenant",
    "account",
    "authKey",
    "session",
    "network",
)


@dataclass
class BudgetConsumeRequest:
    cost: int
    route_key: str
    global_key: Optional[str] = None
    tenant_id: Optional[str] = None
    account_id: Optional[str] = None
    api_key_id: Optional[str] = None
    session_id: Optional[str] = None
    network_key: Optional[str] = None
    tier: Optional[RateLimitTier] = None
    emergency_capacity: Optional[int] = None
    policy: Optional[RouteCostPolicy] = None


@dataclass
class BudgetConsumeResult:
    allowed: bool
    remaining_cost: int
    retry_after_seconds: int = 0
    degraded: bool = False


@dataclass
class RedactedThreatEvent:
    signature: str
    route_template: str
    asn: int = 0
    status_code: Optional[int] = None
    evidence_code: Optional[str] = None
    timestamp: float = 0.0


@dataclass
class ThreatAggregateRecord:
    window_start: str
    window_seconds: int
    signature: str
    route_template: str
    asn: int
    count: int
    samples: List[Dict[str, Any]] = field(default_factory=list)
